### PR TITLE
NEW multiple API updates

### PR DESCRIPTION
**Pull request by c3do**
NEW API ref_ext field for product attributes
NEW API ref_ext field for combinations
NEW API know if an attribute is used when you get it
FIX API result consistency for /products/attributes endpoint
NEW API know fk_product_parent when getting product, if the product is a child
FIX Order API endpoint ::deleteContact because it did not work correctly
FIX Order API endpoint ::postContact results

A parent and a child can be linked according to different combinations of attributes (by entering the reference). These changes are therefore necessary to ensure consistency:
REPLACE fetchByFkProductChild by fetchAllByFkProductChild in ProductCombination.class.php
NEW function getFkProductParentByFkProductChild in ProductCombination.class.php

### CODE REVIEW

File: htdocs/product/class/api_products.class.php [hunk 1]
```diff
@@ -1678,6 +1800,12 @@ private function _fetch($id, $ref = '', $ref_ext = '', $barcode = '', $includest
             throw new RestException(401, 'Access not allowed for login '.DolibarrApiAccess::$user->login);
         }
 
+		$prodcomb = new ProductCombination($this->db);
+        $this->product->fk_product_parent = null;
+        if (($fk_product_parent = $prodcomb->getFkProductParentByFkProductChild($this->product->id)) > 0) {
```
Comment (eldy) [inline htdocs/product/class/api_products.class.php:1805]: This method seems to not exists.

Comment (c3do) [reply htdocs/product/class/api_products.class.php:1805]: The method is added at commit 6bffbc4: https://github.com/Dolibarr/dolibarr/pull/14537/commits/6bffbc43f3f874e6a5c5894663a068f43aece767

**Comment by eldy** [comment]
Can you make "atomic PR". For example adding ref_ext or adding API of contact is not related to change into the ProductCombination.

File: htdocs/product/class/api_products.class.php (hunk 1 above)
Comment (eldy) [reply htdocs/product/class/api_products.class.php:1820]: I did not find the commit you mention is an existing PR. Can you tell me in which PR i can find it ?

File: htdocs/variants/class/ProductCombination.class.php [hunk 2]
```diff
@@ -23,944 +23,983 @@
...
+        }
+    }
+
+    /**
+     * Get fk_product_parent by fk_product_child
+     *
+     * @param int $fk_child Product row id
+     * @return int >0 OK <0 KO
+     */
+    public function getFkProductParentByFkProductChild($fk_child)
```
Comment (c3do) [inline htdocs/variants/class/ProductCombination.class.php:248]: @eldy the method getFkProductParentByFkProductChild is here

**Comment by c3do** [comment]
> Can you make "atomic PR". For example adding ref_ext or adding API of contact is not related to change into the ProductCombination.

i am not sure to understand what do you mean. I preferred to edit everything on the same PR because all the modifications concern the Rest API, and the same files for the majority of the modifications.
afterwards, I have a little difficulty understanding English. It seems to me that you are French, so if you want to ask me more precisely what you would like in French.

**Comment by eldy** [comment]
Having all in same PR is not a good idea.
This is a definition of Atomic PR in the contributing.md file:  
"A PR must be atomic. It means it must contains the lower possible changes for 1 need (1 bug fix or 1 new feature) without breaking usability of code. If a PR can be split into several PRs, it often means your PR is not atomic."

Here, there is problem with getFkProductParentByFkProductChild and it blocks the possibility to merge the other changes (for example changes on the feature to add contacts or the other feature to add ref_ext.

**Comment by c3do** [comment]
> Here, there is problem with getFkProductParentByFkProductChild

ok, i understand and i will make smaller PR next time.

But, i don't see any problem with this method.
It seem to be present in this PR file ProductCombination.class.php https://github.com/Dolibarr/dolibarr/pull/14537#discussion_r482911520

If i do multiple PR now, it will generate a lot of conflicts between them...

**Pull request closed without merging**